cpp/src/parquet/column_reader.cc

```diff
@@ -386,6 +397,16 @@ std::shared_ptr<Page> SerializedPageReader::NextPage() {
       throw ParquetException("Invalid page header");
     }
 
+   // Once we have the header, we will call the skip_page_call_back_ to
+   // determine if we should be skipping this page. If yes, we will advance the
+   // stream to the next page.
+   if(has_skip_page_callback_) {
```

Comment:
What we have done is to let the PageReader be aware of **Offset Index** belong to the pages of the RowGroup. In this way, it can be smart enough to move among pages and handle the skip logic. The prerequisite is https://issues.apache.org/jira/browse/ARROW-10158. If that sounds good, I can pick up ARROW-10158 to contribute our implementation. WDYT? @fatemehp @emkornfield @pitrou

Reply:
> What we have done is to let the PageReader be aware of Offset Index belong to the pages of the RowGroup

Do you have example code here for what you mean?

> I can pick up [ARROW-10158](https://issues.apache.org/jira/browse/ARROW-10158) to contribute our implementation.

Is the implementation compatible with the callback approach?  If you are willing to contribute it, it seems like it would be valuable.

Reply:
Thanks @wgtmac. With this pull request, we would like to specifically support the case where the Page Index is NOT available. We still want to be able to skip pages based on the metadata in that case.

Reply:
> > What we have done is to let the PageReader be aware of Offset Index belong to the pages of the RowGroup
> 
> Do you have example code here for what you mean?
> 
> > I can pick up [ARROW-10158](https://issues.apache.org/jira/browse/ARROW-10158) to contribute our implementation.
> 
> Is the implementation compatible with the callback approach? If you are willing to contribute it, it seems like it would be valuable.

I agree with what @fatemehp has said. My proposal is orthogonal (and compatible) to the current patch. I can contribute the Page Index implementation and then utilize the Offset Index to skip when available.